modules/core/shared/src/main/scala/io/circe/Decoder.scala

```diff
@@ -1433,11 +1443,27 @@ object Decoder
       protected[this] final def parseUnsafe(input: String): ZoneOffset = ZoneOffset.of(input)
     }
 
+  private final case class DeferredDecoder[A](decoder: () => Decoder[A]) extends Decoder[A] {
+    override def apply(c: HCursor): Result[A] = decodeAccumulating(c).toEither.leftMap(_.head)
```

Comment:
I know it seems redundant, but can you reimplement the function you have in `decodeAccumulating` here as well?  The reason is `apply`/non-accumulating should fail FAST/on the first error, while this implementation still runs all the decoders before failing, but only returns the first error.

Reply:
This was a good opportunity to add the caching that was brought up in [typelevel/cats#4414](https://github.com/typelevel/cats/pull/4414)